core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1843,27 +1848,51 @@ void thread_spmc_relinquish(uint64_t cookie)
 		EMSG("Failed to relinquish cookie %#"PRIx64, cookie);
 }
 
-static int set_pages(struct ffa_address_range *regions,
-		     unsigned int num_regions, unsigned int num_pages,
-		     struct mobj_ffa *mf)
+static int add_pages_at(struct ffa_address_range *regions,
+			unsigned int num_regions, unsigned int *idx,
+			struct mobj_ffa *mf)
 {
 	unsigned int n = 0;
-	unsigned int idx = 0;
 
 	for (n = 0; n < num_regions; n++) {
 		unsigned int page_count = READ_ONCE(regions[n].page_count);
 		uint64_t addr = READ_ONCE(regions[n].address);
 
-		if (mobj_ffa_add_pages_at(mf, &idx, addr, page_count))
+		if (mobj_ffa_add_pages_at(mf, idx, addr, page_count))
 			return FFA_INVALID_PARAMETERS;
 	}
 
-	if (idx != num_pages)
-		return FFA_INVALID_PARAMETERS;
-
 	return 0;
 }
 
+static void *spmc_frag_retrieve(uint64_t cookie,
+				uint32_t next_frag_offset,
+				uint32_t *fragment_length)
+{
+	struct thread_smc_args args = {
+		.a0 = FFA_MEM_FRAG_RX,
+		.a1 = (uint32_t)(cookie),        /* handle high addr */
+		.a2 = (uint32_t)(cookie >> 32),  /* handle low addr */
+		.a3 = next_frag_offset,          /* offset */
+		.a4 = my_endpoint_id,            /* sender_vm_id */
```

Comment:
in FF-A v1.1
w4 – Bit[31:16]: Endpoint ID
     – Bit[15:0]: Reserved (MBZ).

maybe it should be sender_id(linux or hyp) << 16 ?

Reply:
• ID of the Owner or Borrower endpoint.
– Bit[31:16]: Endpoint ID.
– Bit[15:0]: Reserved (MBZ).
• Reserved (MBZ) at any virtual FF-A instance.

Reply:
or Reserved (MBZ) at any virtual FF-A instance，and
		.a4 = 0,            /* sender_vm_id */
I'm not really sure.

Reply:
i think 0 is fine, reference:
https://github.com/ARM-software/arm-trusted-firmware/blob/e318411f02f8a9526c97a23d0cdada2f128446e3/services/std_svc/spm/el3_spmc/spmc_shared_mem.c#L1682

i am confused too about the description "Owner or Borrower", i think they are different roles. 
i think sender_id used in optee maybe is not optee partiton id, reference:
https://github.com/OP-TEE/optee_os/blob/a8719249e3e260f172ab11c35314b068c5f25257/core/arch/arm/kernel/thread_spmc.c#L1776

Reply:
Done